src/main/java/org/opensearch/plugin/insights/core/service/grouper/MinMaxHeapQueryGrouper.java

```diff
@@ -0,0 +1,298 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.plugin.insights.core.service.grouper;
+
+import static org.opensearch.plugin.insights.settings.QueryInsightsSettings.TOP_N_QUERIES_MAX_GROUPS;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.PriorityBlockingQueue;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.common.collect.Tuple;
+import org.opensearch.plugin.insights.core.service.store.PriorityQueueTopQueriesStore;
+import org.opensearch.plugin.insights.rules.model.AggregationType;
+import org.opensearch.plugin.insights.rules.model.Attribute;
+import org.opensearch.plugin.insights.rules.model.GroupingType;
+import org.opensearch.plugin.insights.rules.model.MetricType;
+import org.opensearch.plugin.insights.rules.model.SearchQueryRecord;
+import org.opensearch.plugin.insights.settings.QueryInsightsSettings;
+
+/**
+ * Handles grouping of search queries based on the GroupingType for the MetricType
+ * Following algorithm : https://github.com/opensearch-project/OpenSearch/issues/13357#issuecomment-2269706425
+ */
+public class MinMaxHeapQueryGrouper implements QueryGrouper {
+
+    /**
+     * Logger
+     */
+    private static final Logger log = LogManager.getLogger(MinMaxHeapQueryGrouper.class);
+    /**
+     * Grouping type for the current grouping service
+     */
+    private volatile GroupingType groupingType;
+    /**
+     * Metric type for the current grouping service
+     */
+    private MetricType metricType;
+
+    /**
+     * Aggregation type for the current grouping service
+     */
+    private AggregationType aggregationType;
+    /**
+     * Map storing groupingId to Tuple containing Aggregate search query record and boolean.
+     * SearchQueryRecord: Aggregate search query record to store the aggregate of a metric type based on the aggregation type..
+     * Example: Average latency. This query record will be used to store the average latency for multiple query records
+     * in this case.
+     * boolean: True if the aggregate record is in the Top N queries priority query (min heap) and False if the aggregate
+     * record is in the Max Heap
+     */
+    private ConcurrentHashMap<String, Tuple<SearchQueryRecord, Boolean>> groupIdToAggSearchQueryRecord;
+    /**
+     * Min heap to keep track of the Top N query groups and is passed from TopQueriesService as the topQueriesStore
+     */
+    private PriorityQueueTopQueriesStore<SearchQueryRecord> minHeapTopQueriesStore;
```

Comment:
Need to use interface here too.

Reply:
Missed these in the IDE refactoring, let me update